### PR TITLE
PYIC-5799: add journey descriptions

### PR DESCRIPTION
**Pull request by sam803**
## Proposed changes

Adds a name and description to the journey config and puts the states in a 'states' fields

### What changed

Changes to the journey map configurations, the journey map UI (to parse and display the name/desc) and the lambda StateMachineInitializer that loads the configuration from the yaml files. 

Added a new java record to wrap the states in a journey and changes the the initialiser to read this

Screenshot 2024-05-07 at 09 02 38

### Why did it change

In order to keep descriptions and journey map configurations in one place rather than here and on the team manual

- [PYIC-5799](https://govukverify.atlassian.net/browse/PYIC-5799)

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

[PYIC-5799]: https://govukverify.atlassian.net/browse/PYIC-5799?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: journey-map/public/index.html [hunk 1]
```diff
@@ -48,10 +48,15 @@ <h2>Options</h2>
                             <input id="expandNestedJourneysInput" name="expandNestedJourneys" type="checkbox">
                             <span>Show all states within <strong>nested journeys</strong></span>
                         </label>
+                        <label>
+                            <input id="showJourneyDescInput" name="showJourneyDesc" type="checkbox" checked>
+                            <span>Show journey <strong>descriptions</strong></span>
+                        </label>
```
Comment (Joe-Edwards-GDS) [inline journey-map/public/index.html:54]: (I don't know if we need this to be toggleable, but it's fine!)

File: journey-map/public/index.mjs [hunk 2]
```diff
@@ -257,6 +280,10 @@ const setupSearchHandler = () => {
     })
 }
 
+const hideDescription = () => {
+    console.log("hideDescription");
+}
```
Comment (Joe-Edwards-GDS) [inline journey-map/public/index.mjs]: Is this used?

File: lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachineInitializer.java [hunk 3]
```diff
@@ -19,10 +20,13 @@
...
 
 import static com.fasterxml.jackson.core.JsonParser.Feature.STRICT_DUPLICATE_DETECTION;
+import static com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES;
 
 public class StateMachineInitializer {
     private static final ObjectMapper yamlOm =
-            new ObjectMapper(new YAMLFactory()).configure(STRICT_DUPLICATE_DETECTION, true);
+            new ObjectMapper(new YAMLFactory())
+                    .configure(STRICT_DUPLICATE_DETECTION, true)
+                    .configure(FAIL_ON_UNKNOWN_PROPERTIES, false);
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachineInitializer.java]: Are you sure you want this? I think it's good if we fail fast if the YAML is badly formatted...

If you want to scope this more tightly, you can add `@JsonIgnoreProperties(ignoreUnknown = true)` to the class that you want to allow additional properties. (Or just add the properties!)

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml [hunk 4]
```diff
@@ -1,1393 +1,1416 @@
...
+  ALTERNATE_DOC_DL: >-
+    The routes the user is restricted to when they need to mitigate
+    a contra-indicator from the web-based document checking process.
+  ENHANCED_VERIFICATION_F2F_FAIL: >-
+    The routes a user is restricted to when they need to mitigate a
+    contra-indicator from the web-based verification process but have
+    tried, and failed, to mitigate it using face-to-face verification.
+  REPROVE_IDENTITY: >-
+    The route a user must take to reprove their identity after an
+    account intervention.
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml]: I'm not quite sure whether we need to call out each entry point explicitly like this. I'd probably prefer if:
- The description at the journey map level covers the overall aim of the journey - i.e. in this case it's just 'The routes a user can take to prove their identity to a medium confidence level (P2).'
- If we need a detailed description of the entry points themselves, perhaps we should add an optional `description` to those states (but may not be necessary).

This probably makes it easier to format in the journey map UI as well - if this description is only a short sentence!

**Review by Joe-Edwards-GDS — COMMENTED** [review]
A couple of comments, and I think we should also take a bit of care when we merge this...

It's going to cause merge conflicts with anyone doing journey map changes (I think this is mostly just you and @RogVanGDS ), but probably worth notifying the team channel just in case!

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml (hunk 4 above)
Comment (sam803) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml]: Yes I did this initially but changed after finding the [docs](https://team-manual.account.gov.uk/teams/IPV-Core-team/How-IPV-Core-Works/understanding-journey-map/) - I can go back to my previous version

File: lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachineInitializer.java (hunk 3 above)
Comment (sam803) [reply lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachineInitializer.java]: I didn't think you could use that on a record, but I was wrong :)

File: journey-map/public/index.html (hunk 1 above)
Comment (sam803) [reply journey-map/public/index.html:54]: I figured they could take up a fair amount of space, and might be annoying to an expert :)

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml (hunk 4 above)
Comment (Joe-Edwards-GDS) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml]: Cool - happy for you to use your judgement as well.

If you do need updated copy, then Angela said she's happy to produce some (or just adapt what we've already got).

File: journey-map/public/index.html [hunk 5]
```diff
@@ -48,10 +48,15 @@ <h2>Options</h2>
...
                     </fieldset>
                 </form>
                 <div id="nodeInfo">
-                    <h2>Details</h2>
+                    <h2 id="journeyName"></h2>
+                    <h4 id="journeyDesc"></h4>
```
Comment (Joe-Edwards-GDS) [inline journey-map/public/index.html]: Strictly speaking, heading levels shouldn't skip! I think it'd be fine for this to be a plain `<p>`? Perhaps with an `<hr>` underneath (or bottom margin) if you want some visual separation.

TBH the whole thing could use some actual design, but I don't think it'll be top of anyone's priority list :(

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-failed.yaml [hunk 6]
```diff
@@ -1,24 +1,33 @@
...
+description: |
+  A user chose to prove their identity at the Post Office but their attempt failed. For example:
 
-FAILED:
-  events:
-    next:
-      targetState: F2F_FAILED_PAGE
+    - the user might have taken the wrong document
+    - the name they entered in GOV.UK One Login did not match the name on their document
+    - a technical error might have occurred
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-failed.yaml]: Personally I'd condense these into the sentence, but maybe show it to Angela and see what she thinks.

E.g.
> A user chose to prove their identity at the Post Office but their attempt failed. For example: the user might have taken the wrong document or the name they entered did not match the name on their document.

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-pending.yaml [hunk 7]
```diff
@@ -1,76 +1,84 @@
...
 
-PENDING:
-  events:
-    next:
-      targetState: PENDING_PAGE
-      checkFeatureFlag:
-        deleteDetailsEnabled:
-          targetState: PENDING_PAGE_TEST
+  - they have visited the Post Office for in-person checks
+  - their in-person checks have been processed
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-pending.yaml]: (ditto)

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ineligible.yaml [hunk 8]
```diff
@@ -1,77 +1,82 @@
...
+name: Ineligible Journey
 
-INELIGIBLE:
-  events:
-    next:
-      targetState: ANOTHER_WAY_PAGE
-      checkFeatureFlag:
-        ticfCriBeta:
-          targetState: CRI_TICF_BEFORE_ANOTHER_WAY
+description: A user does not have the correct evidence to prove their identity.
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ineligible.yaml]: Perhaps slightly more accurate is
> A user does not have (or chose not to use) a piece of evidence needed to prove their identity.

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/repeat-fraud-check.yaml [hunk 9]
```diff
@@ -1,156 +1,162 @@
+name: Repeat Fraud Check
+description: >-
+  The route a returning user must take whose last fraud
+  check was more than 6 months ago.
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/repeat-fraud-check.yaml]: I wonder if we can avoid specifying 6 months here - for example when we introduce low confidence, those profiles will only expire after 12 months.

Perhaps
> The route a returning user must take if their last fraud check has expired.

**Review by Joe-Edwards-GDS — COMMENTED** [review]
I've left a few suggestions for the descriptions and formatting, but I'm basically happy with this (with the same caveat about merging as before).

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml (hunk 4 above)
Comment (sam803) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml]: I was in two minds really - I preferred the simplicity, but didn't want to miss any information out

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-failed.yaml [hunk 10]
```diff
@@ -1,24 +1,32 @@
-# Entry points
+name: Face to Face Failed
+description: >-
+  A user chose to prove their identity at the Post Office
+  but their attempt failed. For example: the user might
```
Comment (asimms41) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-failed.yaml]:
```suggestion
  but their attempt failed. For example, they might
```

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ineligible.yaml [hunk 11]
```diff
@@ -1,77 +1,84 @@
...
 
-INELIGIBLE:
-  events:
-    next:
-      targetState: ANOTHER_WAY_PAGE
-      checkFeatureFlag:
-        ticfCriBeta:
-          targetState: CRI_TICF_BEFORE_ANOTHER_WAY
+description: >-
+  A user does not have (or chose not to use) a piece
```
Comment (asimms41) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ineligible.yaml]:
```suggestion
  A user does not have or chose not to use a piece
```

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/update-address.yaml [hunk 12]
```diff
@@ -1,120 +1,126 @@
-# Entry points
+name: Update Address
+description: >-
+  The route a returning user must take to update
+  their address
```
Comment (asimms41) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/update-address.yaml]:
```suggestion
  their address.
```

**Review by asimms41 — CHANGES_REQUESTED** [review]
Just some minor text updates. Looks good.

**Review by asimms41 — APPROVED** [review]
LGTM

**Review by Wynndow — APPROVED** [review]

**Pull request merged**